server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3630,15 +3629,19 @@ public void onFailure(Exception e) {
         }
     }
 
-    private class CreateSnapshotTaskExecutor implements ClusterStateTaskExecutor<CreateSnapshotTask> {
+    private class SnapshotTaskExecutor implements ClusterStateTaskExecutor<SnapshotTask> {
         @Override
-        public ClusterState execute(BatchExecutionContext<CreateSnapshotTask> batchExecutionContext) {
+        public ClusterState execute(BatchExecutionContext<SnapshotTask> batchExecutionContext) throws Exception {
             final ClusterState state = batchExecutionContext.initialState();
+            final SnapshotShardsUpdateContext shardsUpdateContext = new SnapshotShardsUpdateContext(batchExecutionContext);
             final SnapshotsInProgress initialSnapshots = state.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY);
-            SnapshotsInProgress snapshotsInProgress = initialSnapshots;
+            SnapshotsInProgress snapshotsInProgress = shardsUpdateContext.computeUpdatedState();
             for (final var taskContext : batchExecutionContext.taskContexts()) {
+                if (taskContext.getTask() instanceof CreateSnapshotTask == false) {
```

Comment:
nit: slight preference to invert this `if` and capture the task directly:

```suggestion
                if (taskContext.getTask() instanceof CreateSnapshotTask task) {
```

Reply:
Just wanted to keep LoC changed lower :D